app/src/main/java/fr/free/nrw/commons/mwapi/ApacheHttpClientMediaWikiApi.java

```diff
@@ -316,13 +358,23 @@ public String revisionsByFilename(String filename) throws IOException {
     }
 
     @Override
-    public boolean existingFile(String fileSha1) throws IOException {
-        return api.action("query")
-                .param("format", "xml")
-                .param("list", "allimages")
-                .param("aisha1", fileSha1)
-                .get()
-                .getNodes("/api/query/allimages/img").size() > 0;
+    public Observable<Boolean> existingFile(final String fileSha1) {
+        return Observable.create(new ObservableOnSubscribe<Boolean>() {
```

Comment:
```java
@Override
public Single<Boolean> existingFile(final String fileSha1) {
    return Single.fromCallable(() -> api.action("query")
            .param("format", "xml")
            .param("list", "allimages")
            .param("aisha1", fileSha1)
            .get()
    ).map(r -> r.getNodes("/api/query/allimages/img").size() > 0);
}
```
Definitely looks a lot nicer than wrapping the API call and manually passing the exception through to `onError()`.  I like it.  If the API call is a one shot (like most are) then `Single` _is_ the way to go, we just need to make sure that we don't leak the `ApiResult` outside of the class.